integrations: Update Slack integration to handle Event API.

PieterCK opened the pull request:
## :page_facing_up: Overview
This PR updates the Slack webhook integration, maintaining its purpose of sending Slack messages to Zulip. The feature for mapping Slack channels is preserved, added better support for different types of messages, and added more tests. The Slack integration now uses Event API instead of the legacy "Outgoing Webhook service".

_Zulip side:_
![image](https://github.com/zulip/zulip/assets/108744694/77c0cab0-eebd-4682-a638-0fc9a5529d32)

_Slack side:_
![image](https://github.com/zulip/zulip/assets/108744694/56e82b56-11be-4cfe-ada8-9a8a6f427d6b)

<details>
<summary>Documentation</summary>

![image](https://github.com/user-attachments/assets/2d46d881-7b0f-43de-aa2a-c8012ec814ba)

</details>

_If used as a Slack Bridge:_

[slackbridgedemo](https://github.com/zulip/zulip/assets/108744694/5852ff60-243e-404b-b0cd-f10b2ea0d32b)





Fixes #30074

---

## Follow up plans
I've documented all potential follow up plan from this PR in #30827 .  

Other than that, now that the Slack integration does callbacks to Slack API, it would also benefit from PR like zulip#10907. I'll cross-link it here to better track relevant integrations that would benefit from that feature,

## Note on testing
If you plan to set up live integration, use ngrok and not ultrahook. Ultrahook overrides all responses from Zulip to Slack, this is problematic when trying to handle Slack's "challenge" handshake for URL verification. If you opt for ngrok, you will have to bypass our subdomain validator (`git grep validate_account_and_subdomain`) by adding an early return.

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

kennethnrk commented:
Looks good other than the test cases. Don't forget to add the documentation.

PieterCK commented:
@zulipbot add "buddy review"

Hello @kennethnrk , I've added a couple more tests and also the doc. Could you please help check those as well? thank you!

PieterCK commented on zerver/webhooks/slack/fixtures/not_yet_supported/message_with_code_block.json:
FOLLOW-UP: files in this folder contain payloads of Slack messages that the integration doesn't currently support. Plans adding support to these types of messages should be relevant for our Slack export conversion tool as well.

PieterCK commented on zerver/webhooks/slack/view.py:
```diff
@@ -1,62 +1,237 @@
+
+            # TODO: As a follow up might want to send a notification message
+            # informing the user about this kind of error in a user-friendly way
+            # instead of logging the error or raising Exception.
```
FOLLOW-UP: Better handling of these kinds of errors could potentially improve UX when trying to debug why their integration doesn't work.

PieterCK replied:
Add back jsonable error

PieterCK commented on zerver/webhooks/slack/view.py:
```diff
@@ -1,62 +1,237 @@
+
+
+def get_slack_channel_info(channel_id: str, token: str) -> SlackAPIResponseT:
+    # TODO: additional improvement would be to cache the response
```
FOLLOW-UP: This integration expects high-volume traffic because we're sending live messages from Slack to Zulip. Doing two callbacks(username & channel) for each message could potentially be problematic for latency especially if this integration is used as part of the Slack Bridge, which is trying to replicate a real-time mirror to Slack. Adding caching might ease the potential latency problem.

PieterCK commented on zerver/webhooks/slack/view.py:
```diff
@@ -1,62 +1,237 @@
+            channel_name = channel_info[1]
+            tokens[iterator] = f"**#{channel_name}**" if channel_name else "**#[private channel]**"
+
+            # TODO: as a follow up, we could probably discuss and decide on a
```
FOLLOW-UP: Both user mention and channel mention in Slack integration don't convert to real mentions, but only to mention-like text. We could potentially add support for channel mapping for mentions.

PieterCK commented:
@zulipbot add "mentor review" 

@sbansal1999 please help review this, thanks!

PieterCK commented:
@zulipbot add "mentor review"

@sbansal1999 this one's ready for your review, thanks

sbansal1999 commented on zerver/webhooks/slack/tests.py:
```diff
@@ -1,99 +1,265 @@
+        expected_message = EXPECTED_MESSAGE.format(user=USER, message=MESSAGE_WITH_NORMAL_TEXT)
         expected_topic_name = "test"
         self.check_webhook(
-            "message_info",
+            "message_with_normal_text",
             expected_topic_name,
-            EXPECTED_MESSAGE,
-            content_type="application/x-www-form-urlencoded",
+            expected_message,
+            content_type="application/json",
```
I think it would be better to use the `expected_topic_name` variable in the `build_webhook_url` function call as well; to indicate that the user-entered topic and the topic in which the message is to sent are the same.

sbansal1999 commented on zerver/webhooks/slack/tests.py:
```diff
@@ -1,99 +1,265 @@
 
     def test_slack_channels_map_to_topics_false_and_user_specified_topic(self) -> None:
-        self.CHANNEL_NAME = "general"
+        # If channels_map_to_topics is false and user specified topic then
+        # override channel to channel mapping and prioritize specified topic.
```
This comment can be removed as this behavior is written in the documentation.

sbansal1999 reviewed (commented):
Thanks @PieterCK for working on this one.

I tested the integration locally using `ngrok` and it worked as expected, just that the setup was not very straightforward. I have left some small comments, overall the code looks good to me.

Also, we would have to move the fixtures from the `not_yet_supported` folder to the folder having other fixtures as this is breaking the integrations dev tools. I wonder what is the current behaviour when a user sends an unsupported message.

sbansal1999 commented on zerver/webhooks/slack/doc.md:
```diff
@@ -1,34 +1,70 @@
 
 See also the [Slack-compatible webhook](/integrations/doc/slack_incoming).
 
+{start_tabs}
```
Probably a new line can be added here to make things consistent.

sbansal1999 commented on zerver/webhooks/slack/doc.md:
```diff
@@ -1,34 +1,70 @@
+   section and make sure "Bot Token Scopes" includes:
+   `channels:read`, `channels:history`, `users:read`, `eomji:read`,
+   `team:read`, `users:read`, `users:read.email`.
+   > *Note: If you worry about the scopes that we're asking, refer to [Slack APIs used](#slack-apis-used) section to see the list of Slack APIs that we use.*
```
You can wrap this line to make it around 70-75 characters per line for readability,

sbansal1999 commented on zerver/webhooks/slack/doc.md:
```diff
@@ -1,34 +1,70 @@
+
+1. Create a new [Slack app][slack_app_link] and open it.
+
+1. Navigate to the "OAuth & Permissions" menu, scroll down to the "Scopes"
```
Replacing the words in quotes with a bold text might be a good idea.

sbansal1999 commented on zerver/webhooks/slack/doc.md:
```diff
@@ -1,34 +1,70 @@
+to. Note that in this case, the channel to channel mapping will be
+prioritized over the channel specified in the URL.
+
+### Slack APIs used
+
+- https://api.slack.com/events/message.channels, this event is used with Event
+   API to get Slack messages.
+- https://api.slack.com/methods/users.info, this end point is used to get
+   the username of each sender. The basic payload only include user_id
+- https://api.slack.com/methods/conversations.info, this API is used to get
+   the channel name of each message. The basic payload only include channel_id
+
+[slack_app_link]: https://api.slack.com/apps
```
This can be replace by something like how's it done for the GitHub integration and just leave the link of the [Events API](https://api.slack.com/apis/events-api) and the [Slack Apps](https://api.slack.com/apps) page.

sbansal1999 reviewed (commented):
Thanks for making the changes. 

Left some comments on the documentation and on the tests as well. I think you forgot to add the tests for the fixtures that are not yet well handled well by us.

Also, you can remove the TODOs from the code and write them in the PR description itself for better visibility.

PieterCK commented:
@sbansal1999 - Thanks for the review! 
Update: 
- Addressed code reviews
- Updated documentation
- Updated PR description
- Opened #30827 to better document follow-up proposals related to this PR.

Sorry for the late response. I've moved the follow up plans to another issue instead of in the PR description as to avoid it getting cluttered too much, does that work with you? let me know if you have any other suggestion and whether the new changes have addressed your concerns!

sbansal1999 commented:
Thanks for making the changes. The code LGTM now. Good work!

>  I've moved the follow up plans to another issue instead of in the PR description as to avoid it getting cluttered too much, does that work with you?

Yes, that sounds perfect.

PieterCK commented:
@zulipbot remove "mentor review"

PieterCK commented:
@sbansal1999 , @alya - I suppose we can proceed to the next round of review for this one?

alya commented:
@laurynmm Are you up for taking a look at this PR next?

laurynmm commented on zerver/webhooks/slack/doc.md:
```diff
@@ -1,34 +1,61 @@
+   enter your webhook url in the **Request URL** field.
+
+1. Still in the same menu, scroll down to the **Subscribe to bot events** and
+   click on **Add Bot User Event**. Finally, select the event `message.channels`.
```
I think we can simplify this a bit too...

```
1. In the same menu, scroll down to the **Subscribe to bot events**
   section, and click on **Add Bot User Event**. Select the
   `message.channels` event.
```

laurynmm commented on zerver/webhooks/slack/doc.md:
```diff
@@ -1,34 +1,61 @@
 
 ![](/static/images/integrations/slack/001.png)
+
+### Channel Mapping
```
Again, let's use the "Confguration options" section header here. And make the points below bulleted.

laurynmm commented on zerver/webhooks/slack/doc.md:
```diff
@@ -1,34 +1,61 @@
+
+### Channel Mapping
+If you'd like to map Slack channels to different topics within the same
+channel, add `&channels_map_to_topics=1` to the end of the URL. Note that
```
Let's clarify Zulip channel here.

laurynmm commented on zerver/webhooks/slack/doc.md:
```diff
@@ -1,34 +1,61 @@
+is specified in the URL, then it will be prioritized over the channel to
+topic mapping.
+
+If you'd like to map Slack channels to different channels, add
```
Let's clarify to different Zulip channels here.

laurynmm commented on zerver/webhooks/slack/doc.md:
```diff
@@ -21,6 +21,9 @@ See also the [Slack-compatible webhook](/integrations/doc/slack_incoming).
    `team:read`, `users:read`, `users:read.email`.
+   > ***Note**: If you worry about the scopes that we're asking, refer to
+   [Related Documentation](#related-documentation) section for more context
+   on the Slack APIs that we use.*
```
Use the warning syntax here. And let's have all the doc updates in one commit.

Also, I'd say "If you're concerned about the scopes specified above, see Slack's Events API documentation for more context." with a direct link here.

laurynmm commented on zerver/webhooks/slack/doc.md:
```diff
@@ -58,4 +61,12 @@ name of each channel is the same as the name of the Slack channel it maps
 
+### Related documentation
+
+- [Slack Event API documentation](https://api.slack.com/apis/events-api)
```
I think Events is plural.

laurynmm commented on zerver/webhooks/slack/doc.md:
```diff
@@ -58,4 +61,12 @@ name of each channel is the same as the name of the Slack channel it maps
+
+- [Slack Event API documentation](https://api.slack.com/apis/events-api)
+
+- [Slack App][slack_app_link]
```
I'd make Apps plural here.

laurynmm commented on zerver/webhooks/slack/view.py:
```diff
@@ -52,8 +54,12 @@ def convert_mentions(text: str) -> str:
-            pass
+            # We convert user mentions into a mention-like message to because
+            # there are no way of mapping between a users Slack and Zulip account
+
```
I'd revise to be not have the empty line after the comment. Also, the comment could be simplified...

`# Convert Slack user mentions to a mention-like syntax since there is no way to map Slack and Zulip users.`

laurynmm reviewed (commented):
@PieterCK - Thanks for doing this update to the new Slack Events API!

Let me know if you have any questions about my review comments, which were mostly about clarifying and tightening up some things.

Also, I'd go back through your commit messages to review for typos and clarity.

PieterCK commented on zerver/webhooks/slack/view.py:
```diff
@@ -27,14 +43,59 @@ def is_bot_message(payload: WildValue) -> bool:
+    # No regex for this as it can be present anywhere in the sentence
+    text = text.replace("<!everyone>", "@**all**")
+    text = text.replace("<!channel>", "@**all**")
+    text = text.replace("<!here>", "@**all**")
```
Yes, a similar follow up plan is already being proposed here: #30827  !

PieterCK commented:
@laurynmm - Thanks for reviewing this one too! 
Update:
- Addressed code reviews
- Revised commit descriptions
- Updated doc screenshot in PR desc

Let me know whether the new changes has addressed your comments or not :+1:

PieterCK commented:
Update: Resolved conflict related to the deprecation of some types from the typing module.

PieterCK commented:
Update: resolved conflict

laurynmm reviewed (commented):
@PieterCK - Thanks for implementing those updates and apologies for not getting back to reviewing this before I was away on vacation.

I had another round of comments, but it's looking really close. Let me know if you have any questions. I went ahead and did some revisions on documentation areas since that seemed quicker and it was mostly tweaking things to read just a bit better.

PieterCK commented on zerver/webhooks/slack/view.py:
```diff
@@ -4,21 +4,27 @@
 
 
+def is_challenge_handshake(payload: WildValue) -> bool:
+    return payload.get("type").tame(check_none_or(check_string)) == "url_verification"
```
I think all of them have it. I'll change it to only `check_string`

PieterCK commented on zerver/webhooks/slack/view.py:
```diff
@@ -27,6 +33,17 @@ def api_slack_webhook(
+            "Integration events",
+            "Successfully verified webhook URL with Slack!",
+        )
+        return json_response(data={"challenge": challenge})
```
oh whoops, yeah it should be `json_success`

PieterCK commented on zerver/webhooks/slack/view.py:
```diff
@@ -44,34 +75,45 @@ def api_slack_webhook(
+        "message_sender": get_sender_name_str(payload),
+    }
+
+    if event_data["type"] == "message":
```
Right, building `event_data` after we confirm the event type to be "message" is better.

laurynmm replied:
Don't think you quite finished this refactor as you've already got an `if` statement for this case above...

 ```
    event_type = event_dict.get("type").tame(check_string)
    if event_type == "message":
```

PieterCK commented:
Updated the PR to address reviews.

@laurynmm - thanks for reviewing again! no worries, hope you enjoyed your vacation :beach_umbrella: . Let me know if there's anything else you'd like to suggest!

laurynmm commented on zerver/webhooks/slack/view.py:
```diff
@@ -1,60 +1,238 @@
+        event_data: dict[str, Any] = {
+            "type": event_dict.get("type").tame(check_string),
+            "text": convert_to_zulip_markdown(raw_text, slack_app_token),
+            "channel": get_channel_name_str(payload, slack_app_token),
```
I wonder if we want to avoid this callback unless there's some sort of valid `channels_map_to_topics` configuration?

```python
if channels_map_to_topics is None:
    ...
elif channels_map_to_topics in VALID_OPTIONS:
    channel = get_channel_name_str ...
    if channels_map_to_topics == VALID_OPTIONS["SHOULD_BE_MAPPED"]:
        ...
    else:
        ...
else:
    JsonableError ...
```

PieterCK replied:
Right, this is a good catch.

laurynmm reviewed (commented):
@PieterCK - Thanks for making that update! With that change, I saw some other ways we could tighten up the view logic around the event data that we're building here and the callback to Slack for the channel name.

I find it's often useful for me to read through my overall PR changes after implementing review feedback because I can sometimes identify opportunities to further tighten/improve my changes once the review feedback is implemented.

Let me know if you have any questions.

PieterCK commented on zerver/webhooks/slack/view.py:
```diff
@@ -1,60 +1,238 @@
+        raw_text = event_dict.get("text", "").tame(check_string)
+
+        event_data: dict[str, Any] = {
+            "type": event_dict.get("type").tame(check_string),
```
Arg, careless mistake. Correct, we don't use event type anymore from this point

laurynmm commented:
@PieterCK - So, I pulled these changes down today and made a few revisions to:

- The order of the commits and the commit messages.
  - The documentation change is the final commit.
  - Made the commit summaries more consistent.
- How/when the event data is dealt with in the main view function.
  - Instead of creating a dict for the data, each value is checked/converted separately.
  - Simplified the logic for the user name and channel name callbacks.
  - Renamed some variables/parameters to avoid abbreviations (e.g., data instead of obj).
  - Moved some of the code comments for readability.

It'd be great if you could review my updates, and let me know if you have any thoughts/feedback.

---

Finally, the first four commits should probably be squashed into one commit. The test changes in the 4th commit are for the code changes in the prior three commits, so it makes sense for them all to be done in one single commit.

That will involve updating that one commit message to encompass the overall changes to the integration to use Slack's Events API vs Slack's deprecated outgoing webhooks.

PieterCK commented on zerver/webhooks/slack/view.py:
```diff
@@ -1,60 +1,187 @@
+        files = convert_raw_file_data(raw_files) if raw_files else []
+
+        raw_text = event_dict.get("text", "").tame(check_string)
+        text = convert_to_zulip_markdown(raw_text) if raw_text else ""
```
added a condition to skip this, to make things consistent

PieterCK commented on zerver/webhooks/slack/view.py:
```diff
@@ -1,60 +1,187 @@
+        text = convert_to_zulip_markdown(raw_text) if raw_text else ""
+
+        sender = event_dict.get("user").tame(check_none_or(check_string))
+        if sender is None:
```
Added a logic to anticipate Slack's integration bot messages. we can add support for this once #31162 is merged, there are a couple of commits that decouple relevant functions from Slack incoming integration to handle this kind of message

PieterCK replied:
added `message_from_slack_integration_bot.json` to document these type of messages

laurynmm replied:
Good catch! Updates look good.

PieterCK commented:
@laurynmm - This looks better. I added a couple more changes and squashed the commits as you suggested. Thanks!

laurynmm commented:
@PieterCK - Great! This is looking good. Going to mark for integration review as I don't have any more feedback at this point.

@timabbott - I'm not sure how we want to handle how these changes would mean things stop working for folks using the current Slack integration that uses Slack's deprecated outgoing webhooks service.

timabbott commented on zerver/webhooks/slack/view.py:
```diff
@@ -1,60 +1,187 @@
-    user_name: str,
-    text: str,
-    channel_name: str,
+    payload: JsonBodyPayload[WildValue],
```
Can we make this support the old Slack API as well by just leaving the old code as an `api_slack_legacy_webhook` function that we call if the payload format matches the old Slack API rather than the new one?

That's how we usually try to handle third-party API changes like this. The other option is to change the endpoint URL to `slack_events` or something, but it's usually nicer.

PieterCK replied:
Oh, yeah, that’s doable. We could create an adapter function that returns a WildValue payload if we receive the old payload. This way, we can just remove the function if we decide to drop support for the old payload later down the line.

PieterCK commented:
I'll work on this tomorrow

PieterCK commented:
@timabbott Sorry this took so long, I was having a hard time trying to make `JsonBodyPayload` optional. Turns out the `slack_incoming` integration faces the same situation ..., so in the end I just have to copy how it’s done there :sweat_smile:. I’ve also updated the doc screenshot in the PR description.

alya commented:
I pushed a commit that should get squashed to update the documentation, and updated the documentation screenshots in the PR description.

laurynmm reviewed (commented):
@PieterCK - Thanks for making the updates to handle the legacy Slack integration! 

My main feedback is that the view and test changes shouldn't be in a final commit, but rather integrated into the first commit where the new Slack API is added. And the documentation changes can all be done in one single commit.

Let me know if you have any questions. It'd be great to keep this moving forward, so if you won't have time to make the above suggested changes, let me know so that I can push those changes here. Thanks!

PieterCK commented on zerver/webhooks/slack/view.py:
```diff
@@ -179,12 +180,43 @@ def api_slack_webhook(
 ) -> HttpResponse:
     topic_name = "Message from Slack"
 
+    if request.content_type == "application/json":
```
Done,

PieterCK commented on zerver/webhooks/slack/doc.md:
```diff
@@ -1,17 +1,32 @@
-public channels!
+Get Zulip notifications from Slack for messages on your team's public channels!
+You can choose to map each **Slack channel** either to a **Zulip channel** or to
+a **Zulip topic**.
```
Done,

PieterCK commented:
Addressed reviews. Let me know if any part could use git diff

laurynmm commented:
@PieterCK - Thanks for making those updates! Some of @alya's documentation edits were lost in your changes though, so I'm pushing the following updates:

- Revisions of my own to @alya's documentation changes, see screenshot below.
- Revisions to the initial commit with the Slack Events API changes:
  - Removes outdated commit reference in commit message.
  - Reorders handling the legacy outgoing webhook case for an early return pattern.
  - Also reorders event type check for a similar pattern.
  - Renamed `handle_send_webhook_message` to `handle_slack_webhook_message` for clarity since that function is mostly handling the `channels_map_to_topics` parameter that's specific to this slack integration.
  - Adjusted some spacing in the view file, and revised some of the code comments.

<details>
<summary>Updated Slack integration doc</summary>

![Screenshot from 2024-09-30 16-06-12](https://github.com/user-attachments/assets/439c7cd3-c082-45a8-a442-0fc4f6117378)
</details>

---

@alya - You may want to do a read through of the documentation updates. Note that for integration docs in general, we've been using `### Related documentation`, so it'd be good to maintain that pattern here. And if we want to change that to `## Related documentation`, then do a bigger sweep for it.

@PieterCK - Let me know if you have any questions about any of the updates that I've pushed.

@timabbott - I think this is ready for you to review at this point as your initial feedback to handle the legacy outgoing webhook case has been addressed.

PieterCK commented:
Ah, sorry about the doc I must have missed that. The changes look good to me, thank you.

timabbott commented on zerver/webhooks/slack/view.py:
```diff
@@ -1,60 +1,226 @@
+        )
+        if slack_usermention_match:
+            # TODO: Callback to Slack API
+            pass
```
@PieterCK can you file an issue to track this follow-up?

timabbott replied:
Never mind, the next commit does this :)

timabbott commented on zerver/webhooks/slack/view.py:
```diff
@@ -1,60 +1,226 @@
+            # a mention isn't triggered for a Zulip channel with the same name.
+            channel_info: list[str] = slack_channelmention_match.group(0).split("|")
+            channel_name = channel_info[1]
+            tokens[iterator] = f"**#{channel_name}**" if channel_name else "**#[private channel]**"
```
Is "unknown Slack channel" a better phrase for this detail? I feel like the alternative has some chance a user might assume it's a Zulip private channel involved here.

PieterCK replied:
Continued the discussion here: [CZO](https://chat.zulip.org/#narrow/channel/127-integrations/topic/Slack.20Webhook.20follow.20up.20feature)

timabbott commented on zerver/webhooks/slack/view.py:
```diff
@@ -1,60 +1,226 @@
+            pass
+        elif slack_channelmention_match:
+            # Convert Slack channel mentions to a mention-like syntax so that
+            # a mention isn't triggered for a Zulip channel with the same name.
```
I don't understand why we're trying to avoid mentioning a Zulip channel with the same name? The only channels a newly slack-imported org will have are ones in this import, and messages and not re-rendered (potentially re-resolving such links) after the import.

PieterCK replied:
Oh, this comment slipped past me.
I was thinking that scenario might occur if the Slack integration is used to connect a different Zulip realm and Slack workspace. It'd probably be confusing if someone from Slack refers to a common channel name like "issues" in their message and it is rendered as referring to the Zulip server's "issue" channel.

timabbott commented on zerver/webhooks/slack/view.py:
```diff
@@ -1,60 +1,226 @@
+    text = convert_markdown_syntax(text, SLACK_BOLD_REGEX, "**")
+    text = convert_markdown_syntax(text, SLACK_STRIKETHROUGH_REGEX, "~~")
+    text = convert_markdown_syntax(text, SLACK_ITALIC_REGEX, "*")
+    return text
```
Hmm, does this one match `convert_slack_formatting` in `slack_message_conversion.py`?

timabbott replied:
This TODO is already in #30827 for tracking.

timabbott commented on zerver/webhooks/slack/view.py:
```diff
@@ -1,60 +1,226 @@
-            content,
-            stream=channel_name,
+            "Integration events",
+            "Successfully verified webhook URL with Slack!",
```
This would be worth posting an example screenshot of in Zulip to discuss if we have better wording we should standardize on for this class of thing.

PieterCK replied:
Continued the discussion here: [CZO](https://chat.zulip.org/#narrow/channel/127-integrations/topic/Notification.20for.20succesful.20.20integration.20setup.2E)

timabbott commented on zerver/webhooks/slack/view.py:
```diff
@@ -204,12 +226,14 @@ def api_slack_webhook(
     if event_type != "message":
         raise UnsupportedWebhookEventTypeError(event_type)
 
+    check_token_access(slack_app_token)
```
Is this call useful? I think it's probably undesirable to do the part of it that makes an API request to check if the token is working every time this webhook is hit; that can add significant latency to the integration. I'm also not sure the set of required scopes would be the same as those needed for the data export tokens, so it probably makes sense to make that an explicit parameter?

Probably we should just call this on the challenge code path, and just have the other code paths that make Slack API requests handle "this token was deactivated" class of exceptions properly.

This can probably be a follow-up, but should be a near-term one.

timabbott commented on zerver/webhooks/slack/doc.md:
```diff
@@ -1,34 +1,79 @@
+
+1. To map each Slack channel to a Zulip topic, [create one channel][2]
+   you'd like to use for Slack notifications. Otherwise, for each public
+   Slack channel, [create a Zulip channel][2] with the same name.
```
Do we support having a non-identity name mapping? Might be important for a channel in a Slack Connect setup where the channel in the third-party org is not what you want in your Zulip org.

timabbott replied:
Added this TODO to #30827 for tracking.

PieterCK replied:
hmm, I don't think we have support for that.

timabbott commented on zerver/webhooks/slack/doc.md:
```diff
@@ -1,34 +1,79 @@
+
+    Otherwise, add `&channels_map_to_topics=0` to the generated URL.
+    Note that any Zulip channel you specified when generating the URL
+    will be ignored in this case.
```
We can use the new `config_options` feature added in d1ff871523841eb15b1f29f67c16c8d66e4bb850 to greatly simplify this documentation, by having the widget generate the required field.

timabbott replied:
Added this TODO to #30827 for tracking.

timabbott commented on zerver/webhooks/slack/doc.md:
```diff
@@ -1,34 +1,79 @@
+
+1. Make sure **Bot Token Scopes** includes `channels:read`,
+   `channels:history`, `users:read`, `emoji:read`, `team:read`,
+   `users:read`, and `users:read.email`.
```
Do we need all of these scopes? I feel like we may want to add a section at the bottom of these docs explaining the reasons for the individual scopes for the paranoid, but maybe it's not worth the effort.

timabbott replied:
Added this TODO to #30827 for tracking.

PieterCK replied:
yes since we need to do callbacks for the channel name and user name

timabbott commented:
Merged, huge thanks for all the work on this @PieterCK! I added several of my review notes to #30827, the ones that are about starting a wording discussion were not migrated, though.

timabbott commented:
This was backported in https://github.com/zulip/zulip/pull/32762.
